zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,276 @@
+            // Do nothing since we are trying to build a data tree in memory, not in actual zk server
+          }
+        });
+
+    DataNode originalNode = dataTree.getNode(targetZNodePath);
+    if (originalNode == null) {
+      return;
+    }
+    Stat originalNodeStat = new Stat();
+    originalNode.copyStat(originalNodeStat);
+
+    LOG.info(
+        "The restored zk data tree is constructed. The highest zxid restored is: " + zxidRestored);
+    LOG.info("Requested path node data: " + Arrays.toString(originalNode.getData()));
+    LOG.info(
+        "Requested path node children: " + Arrays.toString(originalNode.getChildren().toArray()));
+    LOG.info("Requested path node stat: " + originalNodeStat.toString());
+    String requestMsg = "Do you want to restore to this point? Enter \"yes\" or \"no\".";
+    String yesMsg = "Restoring ZooKeeper server...";
+    String noMsg = "Restoration aborted. Nothing has been changed.";
+    if (getUserConfirmation(requestMsg, yesMsg, noMsg)) {
+      traverseNode(zk, dataTree, targetZNodePath, true);
+      printExitMessages();
+    }
+  }
+
+  /**
+   * 1. Restore the node value in zk server;
+   * 2. check if there are deprecated nodes (exist in zk server but not in restored data tree) in its child nodes;
+   * 3. Do the above operations for all of its child nodes in the restored data tree
+   * @param zk A client connection to zk server
+   * @param dataTree The restored zk data tree
+   * @param path The znode path to restore
+   * @param targetNode True if the node being restored is the target znode that user specified;
+   *                   false if it's a child node of the target node
+   */
+  private void traverseNode(ZooKeeper zk, DataTree dataTree, String path, boolean targetNode) {
+    if (!restoreNode(zk, dataTree, path, targetNode) || !restoreRecursively) {
+      // This node is skipped, there's no need to traverse its child nodes
+      return;
+    }
+
+    // Find child nodes in restored data tree
+    DataNode node = dataTree.getNode(path);
+    Set<String> childrenSet = node.getChildren();
+    String[] children = childrenSet.toArray(new String[0]);
+
+    // Find child nodes that's in zk server but not in restored data tree
+    try {
+      List<String> destChildren = zk.getChildren(path, false);
+      for (String destChild : destChildren) {
+        if (!childrenSet.contains(destChild)) {
+          String nodePath = path + "/" + destChild;
+          messages.add(nodePath + ": This node does not exist in the restored data tree.");
+        }
+      }
+    } catch (Exception e) {
+      skipNodeOrStopRestoration(path, e);
+    }
+
+    // Traverse the child nodes of this node
+    for (String child : children) {
+      traverseNode(zk, dataTree, path + "/" + child, false);
+    }
+  }
+
+  /**
+   * Restore the znode in the zk server, currently only doing creation of non-existing nodes,
+   * and only supports persistent nodes
```

Comment:
Creation of non-existing nodes or a single previously non-existent znode? Is this method only dealing with a single ZNode?